### ISSUE TITLE
Docker-desktop + SDS-auth values fails to start

### ISSUE THREAD

**Issue by haf-afa** [istio/istio#18221]
**Bug description**

Just setting up Istio on Docker for Windows locally, fails, with

```
// etc
istio-system   0s          Warning   FailedMount                    pod/istio-ingressgateway-5c4f87768-csdbs       MountVolume.SetUp failed for volume "istio-token" : failed to fetch token: the server could not find the requested resource
istio-system   0s          Warning   FailedMount                    pod/istio-pilot-7576fc8fff-9sttl               MountVolume.SetUp failed for volume "istio-token" : failed to fetch token: the server could not find the requested resource
// etc
```

**Expected behavior**

Since https://istio.io/docs/setup/platform-setup/docker/ doesn't mention any peculiarities and I was under the assumption that Citadel would generate these tokens automatically, it is a bug that istio-token is not generated. Here is one of the many configs that reference the token:

```
      - name: istio-token
        projected:
          sources:
          - serviceAccountToken:
              path: istio-token
              expirationSeconds: 43200
              audience: istio-c
```

in this case from Deployment istio-ingressgateway's spec for istio-proxy.

**Steps to reproduce the bug**

Download istio-1.3.3 to `Infrastructure/istio`, then in a makefile:

```makefile
# From https://istio.io/docs/setup/install/helm/#option-1-install-with-helm-via-helm-template

.PHONY: template
template:
	helm template ../../Infrastructure/istio/helm/istio-init \
		--name istio-init --namespace istio-system \
		> ./istio-init/k8s/base/template.yaml

	# if CNI is re-activated, remember to deploy it also
	helm template ../../Infrastructure/istio/helm/istio \
		--name istio --namespace istio-system \
		--set istio_cni.enabled=false \
		--set sidecarInjectorWebhook.enabled=true \
		--values ../../Infrastructure/istio/helm/istio/values-istio-sds-auth.yaml \
		--set pilot.traceSampling=100.0 \
		--set kiali.enabled=true \
		--set kiali.createDemoSecret=true \
		--set global.tracer.zipkin.address=jaeger-collector.monitoring:9411 \
		--set kiali.dashboard.jaegerURL=jaeger.afa.haf.se \
		> ./istio/k8s/base/template.yaml

# -- TEST aka Docker Desktop

.PHONY: deploy_init_test
deploy_init_test:
	kustomize build ./istio-init/k8s/test | kubectl apply -f -

.PHONY: deploy_istio_test
deploy_istio_test:
	kustomize build ./istio/k8s/test | kubectl apply -f -

deploy_test: template deploy_init_test deploy_istio_test
```

**How was Istio installed?**

Via the above;

 1. Pull down v1.3.3
 2. Template to new k8s cluster on docker for desktop with the above values
 3. Run `make deploy_test` until all CRDs allow instantiations (i.e. until there are no errors reported during installation)
 4. Look at `kubectl get pods --all-namespaces` or `kubectl get all -n istio-system` or `kubectl get events -w --all-namespaces` to see the "failed to fetch token: the server could not find the requested resource"

At this point, restarting Citadel or any other pod in the cluster doesn't magically appear the istio-token secret, so the installation is stuck.

**Environment where bug was observed (cloud vendor, OS, etc)**

- OS: Windows with Docker for Desktop/Windows

TL;DR If you just follow the instruction at https://istio.io/docs/setup/install/helm/#option-1-install-with-helm-via-helm-template but you enable the SDS/auth (strict mTLS) values file, you get this problem.

**Comment by howardjohn** [comment]
@pitlv2109

**Comment by haf-afa** [comment]
Maybe I'm doing something wrong. I just don't know what I'm doing wrong :(

Another data-point, with the default `values.yaml` I still get the istio-token error.

I'm actually trying this for docker-desktop because on GKE I don't get that error, but I have this issue instead: https://stackoverflow.com/questions/58509666/istio-503s-between-public-gateway-and-service albeit with CNI enabled... (I think these are two separate issues)

**Comment by haf-afa** [comment]
https://github.com/istio/istio/issues/17378 is perhaps related

**Comment by pitlv2109** [comment]
Did #17378 fix your issue (using the flags and their values)?

GKE >= 1.12 has trustworthy jwt installed by default. OSS Kubernetes does not.

We didn't want to include flag names and values in the blog because we didn't want to create yet another source of truth (that could diverge from Kubernetes').

**Comment by haf** [comment]
I’m reluctant to change things like this unless it is all automated since I work in a team. 

I’ll give it a shot tomorrow. Do you have sample values you could post? What do you use yourself on docket desktop?

**Comment by pitlv2109** [comment]
I think the env is irrelevant here. If you're running with out of the box OSS Kubernetes, feel free to use the flags we use here https://github.com/istio/istio/tree/master/prow/config for Istio testing.

**Comment by haf** [comment]
Yes, we have both a development environment and a GKE environment; but Docker for Desktop does not expose kubeadm configuration (that I can find), so applying the `kubeadmConfigPatches` seems like a hard sell to me. (I can't apply it with kustomize and there's no configuration section of Docker for Desktop available)

IMO if you require your users to change the kubernetes API server from its defaults, you've moved beyond those same users and have failed at the open source bit of the project. I really appreciate the work you're putting in, but it has to be usable, even if "less secure on the development laptop", which would be the effect of not requiring "trustworthy JWTs". Don't you agree?

**Comment by pitlv2109** [comment]
I find it strange that you cannot configure your k8s installation on Docker for Desktop. I was able to install K8s using KinD with https://docs.docker.com/install/linux/docker-ce/ubuntu/. I'm not an expert at installation k8s, @howardjohn might know more, but again, I feel like there should be a way.

**Comment by howardjohn** [comment]
@haf-afa SDS is not a required feature of Istio, feel free to disable it or use a kubernetes deployment that supports the enhanced security features. I like https://kind.sigs.k8s.io

**Comment by haf** [comment]
@howardjohn It's not just SDS; just starting Istio fails.

**Comment by howardjohn** [comment]
Starting Istio will work if you do not enable SDS. 

My understanding of this is that Istio offers SDS, for users who want enhanced security, or the default file mounted certs. Without this trustworthy jwt feature that requires kubernetes API flags, there are security issues with how SDS behaves. So it doesn't really make sense to use a security feature that is not secure... You do not *need* to use SDS, but if you want to use an optional feature, then yes, you will need to configure Kubernetes to support it

**Comment by haf** [comment]
No it does not start if I use only the defaults. See the linked gist. @howardjohn

**Comment by haf** [comment]
To reply to your argument. It makes sense for any type of production, but sometimes there is a lot of value in having little to no variance in how systems are configured. Take the large outages in AWS stemming from config changes.

Or take another side; developers test against mocks and stubs; but if one takes your argument at face value that would be useless. But in reality it’s not; having mocks let’s you try the integration of your component with the rest. As would starting the same services have such value.

**Comment by howardjohn** [comment]
> No it does not start if I use only the defaults. There is still the missing token-error then. 1.2.8 starts but not 1.3.3.

This sounds like a real bug, this should not he happening if sds is disabled. 

Your point does make a lot of sense though. I think it's likely when we made this change we didn't fully realize the UX implications (although I was not involved in the decision, and I'm not terribly familiar with the technical details here).

Also, not to say your mocking point is not valid, but ideally in development you would be running with a similar kubernetes env, with the same API server flags, in which case this would work.

@pitlv2109 is it technically feasible or desirable to add a "insecure" or "development" mode for this?

**Comment by pitlv2109** [comment]
> No it does not start if I use only the defaults. There is still the missing token-error then. 1.2.8 starts but not 1.3.3.

This should not happen. I'll double check, but I don't remember anyone reporting any issues when 1.3 was out. There was no trustworthy jwt-related changes from 1.3 to 1.3.3.

> @pitlv2109 is it technically feasible or desirable to add a "insecure" or "development" mode for this?

If you're talking about whether one can use legacy jwt with SDS, the answer is no. The original design supported both types of jwt, but there was a question regarding the overhead of maintaining support for both types of jwt. It went to TOC and eventually TOC approved to have one jwt only.

**Comment by pitlv2109** [comment]
I accidentally edited John's comment instead of posting a new one for me 🤦

I just tried creating a OSS Kubernetes cluster using KinD and deployed Istio 1.3.3 without any problem. 

```
NAME                                      READY   STATUS      RESTARTS   AGE
grafana-59d57c5c56-7bk5k                  1/1     Running     0          5m4s
istio-citadel-66f699cf68-vh798            1/1     Running     0          5m4s
istio-egressgateway-559744f57f-4982b      1/1     Running     0          5m4s
istio-galley-f6bf79455-84lp8              1/1     Running     0          5m4s
istio-grafana-post-install-1.3.3-flmjj    0/1     Completed   0          5m5s
istio-ingressgateway-6785fb8874-2sckm     1/1     Running     0          5m4s
istio-pilot-7c464fff87-rccdb              2/2     Running     0          5m4s
istio-policy-b67f89987-lcv75              2/2     Running     2          5m4s
istio-security-post-install-1.3.3-vhsj9   0/1     Completed   0          5m5s
istio-sidecar-injector-d8856c48f-dks8b    1/1     Running     0          5m4s
istio-telemetry-fbbfbf975-vnrgk           2/2     Running     2          5m4s
istio-tracing-6bbdc67d6c-xkkj4            1/1     Running     0          5m3s
kiali-8c9d6fbf6-hvwhv                     1/1     Running     0          5m4s
prometheus-7d7b9f7844-zhrrf               1/1     Running     0          5m4s
```

**Comment by haf-afa** [comment]
> If you're talking about whether one can use legacy jwt with SDS, the answer is no. The original design supported both types of jwt, but there was a question regarding the overhead of maintaining support for both types of jwt. It went to TOC and eventually TOC approved to have one jwt only.

So if this is the case, what is the recommended configuration to have test/dev parity? Do you recommend against using Docker Desktop? In that case, how does binding to low ports (port 443, 80) work on Windows with KindD (because minikube is not a good solution, with its constantly changing IP)?

> I just tried creating an OSS Kubernetes cluster using KinD and deployed Istio 1.3.3 (no SDS) without any problem. 

Did you try starting Istio 1.3.3 without SDS on D/D?

**Comment by haf** [comment]
>  it's a Kubernetes feature

Indeed; which is why it would be awesome if it's easy for people to use this new feature in the default k8s config of docker's installation package. :)

> What I can do is ask the TOC to consider your use case to see if it makes sense to keep supporting legacy JWT for dev purposes.

👍

**Comment by pitlv2109** [comment]
In the meantime, could you show me the error logs when you tried deploying Istio 1.3.3 without SDS? Also run `kubectl -n istio-system get pod` to see if there is a node-agent pod.

**Comment by haf-afa** [comment]
@pitlv2109 Absolutely, I'm going to loop back around to it in a few days; in particular if I can get a good state/invariant test setup going, like asked about in #18225

**Comment by Darwiner** [comment]
For the record (in case anyone else winds up looking around at this, although I might be mistaken?), it seems the Kubernetes' link that you refer to (https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection), itself, is already outdated and doesn't reflect the right information.

This `--service-account-api-audiences` flag seems to have been deprecated/removed/changed (https://github.com/kubernetes/kubernetes/issues/70104) in kube-apiserver and is now `--api-audiences` (https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/).

**Comment by pitlv2109** [comment]
@Darwiner Thank you for bringing this up. Yes that what I found out when I was experimenting with trustworthy JWT. In the blog post, it says "You may also want to refer to the api-server page for the most up-to-date flag names". We didn't want to create another yet source of truth, since this is a K8s feature.

**Comment by haf-afa** [comment]
Here's a stand-alone repro that fails for 1.2.8 and 1.3.4: https://github.com/haf-afa/istio-bugs

Repro: make istio_1_3_4 or make istio_1_2_8

**Comment by haf-afa** [comment]
We've migrated away from SDS/CertManager/Vault+Istio/Ingress Gateway, due to this; the UX made it impossible to use. I could let the issue be open or we can close it. Thank you for your help making me understand the choices you've made and how it all fits together.

**Comment by cioc** [comment]
The blog post and documentation are both unclear as to how to alleviate this issue.  There doesn't seem to be any source of truth on how to proceed.  As @haf pointed out, this is a blocker for adoption of istio.

**Comment by pitlv2109** [comment]
> The blog post and documentation are both unclear as to how to alleviate this issue. There doesn't seem to be any source of truth on how to proceed. As @haf pointed out, this is a blocker for adoption of istio.

If you go to [one](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/bound-service-account-tokens.md#footnotes) of the k8s links from the blog post, you'll see it mentions new api-server flags. K8s also has document about how to set api-server flags. [These](https://github.com/istio/istio/tree/master/prow/config) are the configs that we use depending on k8s version.

**Comment by haf-afa** [comment]
The flags are documented but not how to find out their values. @pitlv2109

**Comment by haf** [comment]
Docker-desktop as of 2020-02-01 does still not pass these flags:

```
    - --secure-port=6443
    - --service-account-key-file=/run/config/pki/sa.pub
    - --service-cluster-ip-range=10.96.0.0/12
```

Did you create that task about setting it up?

**Comment by howardjohn** [comment]
In the next release we are going to support SDS without those flags passes

**Comment by haf-afa** [comment]
@howardjohn So we've moved onto the next release now, Istio 1.5.2 and now it would seem that it's even impossible to turn off SDS locally, thus stopping us from testing istio in docker-desktop.

> istio-system   istiod-5f75f97794-982x5                  0/1     ContainerCreating   0          23s

**Comment by howardjohn** [comment]
please take a look at https://preliminary.istio.io/docs/ops/best-practices/security/#configure-third-party-service-account-tokens. depending on how you install it may require manually setting the flag

**Comment by haf** [comment]
I read the docs on `docker-desktop` for v1.5, are those not up to date?

**Comment by haf** [comment]
Why can't istiod just check what environment it's running in and do the right thing??

**Comment by howardjohn** [comment]
> and I am, but using template files. Does this sentence not apply then (Bug 2)?

This only applies when using `manifest apply`. It is not possible when doing `generate`, because that command does not connect to the kubernetes cluster and therefor cannot do automatic detection.

> This doesn't work on docker-desktop (Bug 3).

By "doesn't work" do you mean it returns nothing? That means it is working -- and indicates your cluster does not support the feature. The doc here could be a little more clear

**Comment by haf** [comment]
> This only applies when using manifest apply. It is not possible when doing generate, because that command does not connect to the kubernetes cluster and therefor cannot do automatic detection.

I understand, but I'm using this thing; code as config, where we place all our config in our repositories, to make deployments repeatable. Having the **templates** be smart is not in our best interest, and we're not going to start versioning the `istioctl` in our git repo.

We want this to be smart:

- the software we run inside k8s, and if that's not possible
- the operator that runs the software, and if that's not possible,
- the config-as-code via kustomize, and if that's not possible,
- I don't use the software

How do I fix this with a "recommended" solution?

> By "doesn't work" do you mean it returns nothing? That means it is working -- and indicates your cluster does not support the feature. The doc here could be a little more clear

Right, so that means I must use "first-party" JWTs then? Can't Istio know this?

**Comment by howardjohn** [comment]
What you are asking for is not possible (or I couldn't figure it out anyways - suggestions welcome). We need to determine the jwt policy at install time, NOT at runtime, because it's part of the deployment spec. That is why we do it in istioctl, and not part of pilot for example. So in your case you will need to explicitly set the option as you mentioned

**Comment by haf** [comment]
@howardjohn What I'm saying is that perhaps it should be part of the runtime? I'm not familiar with the Istio code-base, but is there something architecturally that prohibits doing it at runtime?

We've run into another problem with first-party-jwts with Elastic ECK: `{"level":"info","time":"2020-04-29T02:31:08.581859Z","msg":"JWT policy is first-party-jwt"}`; but my colleagues will be joining in and explain that problem.

**Comment by howardjohn** [comment]
@haf its part of the pod spec. If we try to use 3p jwt then the pod won't even start so we cannot determine at runtime. See https://github.com/kubernetes/kubernetes/issues/86827

For the other issue, maybe you are setting automountServiceAccountToken to false?

**Comment by flerpo** [comment]
Hi, automountServiceAccountToken did the trick! Thanks

**Comment by haf** [comment]
```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: apm
spec:
  nodeSets:
  - name: violet
    count: 1
    # snip
    podTemplate:
      spec:
        automountServiceAccountToken: true
```

**Comment by haf-afa** [comment]
So 
- [ ] bug2
- [ ] bug3 of https://github.com/istio/istio/issues/18221#issuecomment-620618338 needs to be fixed together with
- [ ] documenting what "automounting service account token" means: https://github.com/istio/istio/issues/18221#issuecomment-620618338 
- [ ] as well as documenting how to pass flags to `generate` (`istioctl manifest generate --set values.global.jwtPolicy=first-party-jwt --filename istio-demo.yaml > k8s-1.5/istio-test/template.yaml`)

for this to be solved. It's a docs bug now.
